src/library_wasi.js

```diff
@@ -261,17 +260,41 @@ var WasiLibrary = {
     if (printCharBuffers[1].length) printChar(1, {{{ charCode("\n") }}});
     if (printCharBuffers[2].length) printChar(2, {{{ charCode("\n") }}});
   },
-  fd_write__deps: ['$flush_NO_FILESYSTEM', '$printChar'],
-  fd_write__postset: () => addAtExit('flush_NO_FILESYSTEM()'),
+#if SYSCALLS_REQUIRE_FILESYSTEM
+  fd_write__postset: () => addAtExit('if(ENVIRONMENT_IS_WASM_WORKER) flush_NO_FILESYSTEM()'),
```

Comment:
I don't think this is needed is the `addAtExit` mechanism is already limited to just the main thread.